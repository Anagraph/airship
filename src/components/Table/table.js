import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

/*
  <Table>
    <Table.Header>
      <Table.Row>
        <Table.HeaderCell>Header</Table.HeaderCell>
        <Table.HeaderCell>Header</Table.HeaderCell>
        <Table.HeaderCell>Header</Table.HeaderCell>
      </Table.Row>
    </Table.Header>

    <Table.Body>
      <Table.Row>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
        <Table.Cell>Cell</Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
*/

const StyledTable = styled.table`
  border-collapse: collapse;
  border-spacing: 0;
  border: 1px solid #d1d5d7;
  width: 100%;
  font: 400 12px/20px 'Roboto';
  color: #2c2c2c;

  th {
    font-weight: 600;
    text-align: left;
    background: #e2e6e3;
  }

  th,
  td {
    padding: 8px 12px;
    border-bottom: 1px solid #d1d5d7;
    border-right: ${props => (props.lined ? '1px solid #d1d5d7' : null)};
  }

  tr:last-child td {
    border-bottom: 0;
  }

  tr:hover,
  td:hover {
    background: rgba(71, 219, 153, 0.16);
  }

  th:last-child,
  td:last-child {
    border-right: ${props => (props.lined ? 0 : null)};
  }
`;

const Header = ({ children }) => <thead>{children}</thead>;

Header.propTypes = {
  children: PropTypes.node,
};

const HeaderCell = ({ children, ...props }) => <th {...props}>{children}</th>;

HeaderCell.propTypes = {
  children: PropTypes.node,
  colSpan: PropTypes.number,
  rowSpan: PropTypes.number,
};

const Body = ({ children }) => <tbody>{children}</tbody>;

Body.propTypes = {
  children: PropTypes.node,
};

const Row = ({ children }) => <tr>{children}</tr>;

Row.propTypes = {
  children: PropTypes.node,
};

const Cell = ({ children, ...props }) => <td {...props}>{children}</td>;

Cell.propTypes = {
  children: PropTypes.node,
  colSpan: PropTypes.number,
  rowSpan: PropTypes.number,
};

const Table = ({ children, lined }) => (
  <React.Fragment>
    <StyledTable lined={lined}>{children}</StyledTable>
  </React.Fragment>
);

Table.Header = Header;
Table.HeaderCell = HeaderCell;
Table.Body = Body;
Table.Row = Row;
Table.Cell = Cell;

Table.propTypes = {
  children: PropTypes.node,
  lined: PropTypes.bool,
};

export default Table;
