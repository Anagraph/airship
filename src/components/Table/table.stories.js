import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import Table from './table';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  ui04: '#607D8B',
  ui03: '#607D8B',
  brand03: '#bae7fa',
};

storiesOf('Table', module)
  .add('Default', () => (
    <Table>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>Header</Table.HeaderCell>
          <Table.HeaderCell>Header</Table.HeaderCell>
          <Table.HeaderCell>Header</Table.HeaderCell>
        </Table.Row>
      </Table.Header>

      <Table.Body>
        <Table.Row>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
          <Table.Cell>Cell</Table.Cell>
        </Table.Row>
      </Table.Body>
    </Table>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <Table>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Header</Table.HeaderCell>
            <Table.HeaderCell>Header</Table.HeaderCell>
            <Table.HeaderCell>Header</Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    </ThemeProvider>
  ));
