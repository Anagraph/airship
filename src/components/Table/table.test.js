import React from 'react';
import renderer from 'react-test-renderer';
import Table from './table';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(
      <Table>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Header</Table.HeaderCell>
            <Table.HeaderCell>Header</Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with lined param', () => {
    const component = renderer.create(
      <Table lined>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Header</Table.HeaderCell>
            <Table.HeaderCell>Header</Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with colSpan and rowSpan params for cells', () => {
    const component = renderer.create(
      <Table lined>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell colSpan={2}>Header</Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          <Table.Row>
            <Table.Cell rowSpan={2}>Cell</Table.Cell>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Cell</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
