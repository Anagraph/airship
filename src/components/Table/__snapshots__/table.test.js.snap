// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders with colSpan and rowSpan params for cells 1`] = `
.c0 {
  border-collapse: collapse;
  border-spacing: 0;
  border: 1px solid #d1d5d7;
  width: 100%;
  font: 400 12px/20px 'Roboto';
  color: #2c2c2c;
}

.c0 th {
  font-weight: 600;
  text-align: left;
  background: #e2e6e3;
}

.c0 th,
.c0 td {
  padding: 8px 12px;
  border-bottom: 1px solid #d1d5d7;
  border-right: 1px solid #d1d5d7;
}

.c0 tr:last-child td {
  border-bottom: 0;
}

.c0 tr:hover,
.c0 td:hover {
  background: rgba(71,219,153,0.16);
}

.c0 th:last-child,
.c0 td:last-child {
  border-right: 0;
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th
        colSpan={2}
      >
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        rowSpan={2}
      >
        Cell
      </td>
      <td>
        Cell
      </td>
    </tr>
    <tr>
      <td>
        Cell
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`render renders with lined param 1`] = `
.c0 {
  border-collapse: collapse;
  border-spacing: 0;
  border: 1px solid #d1d5d7;
  width: 100%;
  font: 400 12px/20px 'Roboto';
  color: #2c2c2c;
}

.c0 th {
  font-weight: 600;
  text-align: left;
  background: #e2e6e3;
}

.c0 th,
.c0 td {
  padding: 8px 12px;
  border-bottom: 1px solid #d1d5d7;
  border-right: 1px solid #d1d5d7;
}

.c0 tr:last-child td {
  border-bottom: 0;
}

.c0 tr:hover,
.c0 td:hover {
  background: rgba(71,219,153,0.16);
}

.c0 th:last-child,
.c0 td:last-child {
  border-right: 0;
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th>
        Header
      </th>
      <th>
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Cell
      </td>
      <td>
        Cell
      </td>
    </tr>
    <tr>
      <td>
        Cell
      </td>
      <td>
        Cell
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`render renders without crashing 1`] = `
.c0 {
  border-collapse: collapse;
  border-spacing: 0;
  border: 1px solid #d1d5d7;
  width: 100%;
  font: 400 12px/20px 'Roboto';
  color: #2c2c2c;
}

.c0 th {
  font-weight: 600;
  text-align: left;
  background: #e2e6e3;
}

.c0 th,
.c0 td {
  padding: 8px 12px;
  border-bottom: 1px solid #d1d5d7;
}

.c0 tr:last-child td {
  border-bottom: 0;
}

.c0 tr:hover,
.c0 td:hover {
  background: rgba(71,219,153,0.16);
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th>
        Header
      </th>
      <th>
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Cell
      </td>
      <td>
        Cell
      </td>
    </tr>
    <tr>
      <td>
        Cell
      </td>
      <td>
        Cell
      </td>
    </tr>
  </tbody>
</table>
`;
